Separate jars for XML, Parser Combinators. Use released JLine.

adriaanm opened the pull request:
The actual refactoring had already been done.
This changes the source directory layout and the produced jars.

Review by @gkossakowski

/cc @huitseeker, @harrah, @jsuereth

adriaanm committed:
Spin off src/library/scala/xml to src/xml/scala/xml.

Summary:
  - Remove the last vestiges of xml from Predef and Contexts.
  - Change build to compile scala.xml to scala-xml.jar.
  - Deploy scala-xml module to maven.
  - Update partest accordingly.

Note:
An older compiler cannot use the new standard library to
compile projects that use XML. Thus, skipping locker will
break the build until we use 2.11.0-M4 for STARR.

In the future build process, where we drop locker,
we would have to release a milestone that supports the old
and the new approach to xml. As soon as we'd be using that
new milestone for starr, we could drop support for the old approach.

adriaanm committed:
Unfork jline: use vanilla jline 2.11 as a dependency.

Notes:
  - no longer specifying terminal by class name in scripts (using 'unix')
  - jline doesn't need a separate jansi dependency;
    it includes its own version according to:
    http://mvnrepository.com/artifact/jline/jline/2.11

adriaanm committed:
Spin off parser combinators to scala-parser-combinators.jar.

adriaanm committed:
No more duplication in maven-deploy.xml.

I just couldn't stand the incredible mess in there anymore. More cleanup to come.
For now, suffice it to say you need only add one line per new module.

soc commented:
Anyway, LVGTM!

dragos commented on src/eclipse/reflect/.classpath:
```diff
@@ -1,8 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <classpath>
 	<classpathentry kind="src" path="reflect"/>
-	<classpathentry combineaccessrules="false" kind="src" path="/scala-library"/>
```
In this way the classpath _won't_ contain the library that was just built. Instead it will use the Eclipse built-in one, and maybe you'll need changes in both at the same time (reflect and library are still tightly coupled). Just something to keep in mind.

adriaanm replied:
Good point, this was a mistake.

adriaanm replied:
Reflect should be built against the library. The projects currently built during quick but not during locker should be built using the Scala version in the plugin.

soc commented:
I have no clue about the Eclipse stuff, but the rest LGTM.

harrah commented:
This LGTM.  We'd need coordination for nightly builds against 2.11 if we aren't ok with transient breakage in projects that use xml or parser combinators until builds are updated.

adriaanm commented:
@harrah, thanks. Yes, I'd like to minimize that breakage, but need to merge this into M4, which we're releasing next week. Do you have some time next week to update your build? /cc @dragos, @huitseeker for IDE

adriaanm committed:
Add meta-information for dbuild.

The next version of [dbuild](http://typesafehub.github.io/distributed-build/0.5.3/index.html)
will parse `dbuild-meta.json` to determine which jars are produced by a Scala build.

This way we can modularize without changing dbuild itself.

Yes, I know `dbuild-meta.json` should be generated during the build.
However, given the state of our build.xml, I think this is pointless.

My goal is to generate build.xml, dbuild-meta.json and Eclipse projects
from a higher-level description of our build. Baby steps...

Including improvements by @cunei:
  - Removing outdated field "uri" from ExtractedBuildMeta
  - Changed "partest" to "scala-partest" (the actual jar name)

adriaanm commented:
/cc @cunei for caa3989

soc commented:
Weird. Looks like it can't find `Profiler.java`, although it's supposed to be in the same package.

harrah commented:
@adriaanm yes.  The changes should be straightforward, but require switching on the Scala version.  So, once they are added, the build won't work on older 2.11 versions.  I'm just going to look at >=2.11.x vs <= 2.10.x and not try to get the exact nightly.  I assume that is all ok.

adriaanm commented:
@harrah, that's fine with me -- I'm not hearing anything from the IDE team about his PR, so I'll double check with them.

adriaanm commented:
> Weird. Looks like it can't find Profiler.java, although it's supposed to be in the same package.

It's because java and scala are compiled separately and partest only saw it's own class files through its dependency on scalap. Fixed now.

gkossakowski commented:
I reviewed all commits pushed so far. Great work!

I think there are two outstanding (small) tasks:
- [x] improved commit message in fa31d3c that explains why it's ok to get rid of migration code before M4 is released
- [x] [determine](https://github.com/adriaanm/scala/commit/2391c20e56cec91cf327c0e42fddeefe2e26c963#src-build-maven-maven-deploy-xml-P81) whether `$` is missing in 2391c20

adriaanm committed:
Updated eclipse project files.

Set everything up so that soon-to-be independent modules are
compiled with the Eclipse plugin's compiler & library (2.11.0-M3 currently).

Most projects still compile with 2.11.0-M3, but
partest will need a nightly build of 2.11.

adriaanm commented:
Rebased. Thanks for the reviews everyone!

scala-jenkins commented:
IDE validation running at https://scala-webapps.epfl.ch/jenkins/job/pr-scala/87/

scala-jenkins commented:
IDE validation running at https://scala-webapps.epfl.ch/jenkins/job/pr-scala/88/

scala-jenkins commented:
IDE validation running at https://scala-webapps.epfl.ch/jenkins/job/pr-scala/89/

scala-jenkins commented:
IDE validation running at https://scala-webapps.epfl.ch/jenkins/job/pr-scala/90/

scala-jenkins commented:
IDE validation running at https://scala-webapps.epfl.ch/jenkins/job/pr-scala/92/

scala-jenkins commented:
IDE validation running at https://scala-webapps.epfl.ch/jenkins/job/pr-scala/91/

gkossakowski commented:
LGTM.

Great work! We are celebrating tomorrow with :pizza: and :cocktail: :cocktail:. :-)

adriaanm commented:
Yay! Note that this breaks the eclipse build as SBinary apparently uses xml, so it'll need scala-xml.jar on its classpath

harrah commented:
It is a Format instance for XML that can be dropped for sbinary 0.5.0 to keep its dependencies to just scala-library.

adriaanm commented:
Ok, [dropped the offending lines in my fork](https://github.com/adriaanm/sbinary/commit/9bc0c1123d92bb60427217caf0072b8fa28ffcf8).

adriaanm committed:
scaladoc needs xml and parser-combinators

Concretely, update scala-compiler's pom
to list scala-xml and scala-parser-combinators
on behalf of scaladoc.

NOTE: when spinning off scaladoc, move dependencies to its own pom

adriaanm commented:
To get SBinary to compile: https://github.com/adriaanm/sbinary/commit/7ce59f70d8ea9124cc703af52430be1596e7c9d4)
SBT: https://github.com/adriaanm/sbt/commit/de3664b4e3a1de0859ff1353c2043285cdf711a6

Now only scala-refactoring test failures remain, as some tests depend on parsers and xml
https://scala-webapps.epfl.ch/jenkins/job/pr-scala-integrate-ide/344/console

scala-jenkins commented:
IDE validation running at https://scala-webapps.epfl.ch/jenkins/job/pr-scala/106/
https://scala-webapps.epfl.ch/jenkins/job/pr-scala/87/

scala-jenkins commented:
IDE validation running at https://scala-webapps.epfl.ch/jenkins/job/pr-scala/138/
https://scala-webapps.epfl.ch/jenkins/job/pr-scala/88/

scala-jenkins commented:
IDE validation running at https://scala-webapps.epfl.ch/jenkins/job/pr-scala/133/
https://scala-webapps.epfl.ch/jenkins/job/pr-scala/89/

scala-jenkins commented:
IDE validation running at https://scala-webapps.epfl.ch/jenkins/job/pr-scala/128/
https://scala-webapps.epfl.ch/jenkins/job/pr-scala/90/

scala-jenkins commented:
IDE validation running at https://scala-webapps.epfl.ch/jenkins/job/pr-scala/122/
https://scala-webapps.epfl.ch/jenkins/job/pr-scala/92/

scala-jenkins commented:
IDE validation running at https://scala-webapps.epfl.ch/jenkins/job/pr-scala/117/
https://scala-webapps.epfl.ch/jenkins/job/pr-scala/97/
https://scala-webapps.epfl.ch/jenkins/job/pr-scala/91/

scala-jenkins commented:
IDE validation running at https://scala-webapps.epfl.ch/jenkins/job/pr-scala/147/
https://scala-webapps.epfl.ch/jenkins/job/pr-scala/98/

OlegYch commented:
note that new jline breaks repl on windows https://github.com/sbt/sbt/issues/763
